app/src/main/java/org/kiwix/kiwixmobile/localFileTransfer/PeerGroupHandshake.kt

```diff
@@ -0,0 +1,119 @@
+import java.net.Socket
+
+/**
+ * Helper class for the local file sharing module.
+ *
+ * Once two peer devices are connected through wifi direct, this task is executed to perform a
+ * handshake between the two devices. The purpose of the handshake is to allow the file sending
+ * device to obtain the IP address of the file receiving device (When the file sending device
+ * is the wifi direct group owner, it doesn't have the IP address of the peer device by default).
+ *
+ * After obtaining the IP address, the sender also shares metadata regarding the file transfer
+ * (no. of files & their names) with the receiver.
+ */
+abstract class PeerGroupHandshake(private var groupInfo: WifiP2pInfo) {
+  private val HANDSHAKE_MESSAGE = "Request Kiwix File Sharing"
+  suspend fun handshake(): InetAddress? = withContext(Dispatchers.IO) {
+    if (BuildConfig.DEBUG) {
+      Log.d(TAG, "Handshake in progress")
+    }
+    when {
+      groupInfo.groupFormed && groupInfo.isGroupOwner && isActive ->
+        readHandshakeAndExchangeMetaData()
+      groupInfo.groupFormed && isActive ->  // && !groupInfo.isGroupOwner
+        writeHandshakeAndExchangeMetaData()
+      else -> null
+    }
+  }
+
+  private fun writeHandshakeAndExchangeMetaData(): InetAddress? {
+    return try {
+      Socket().use { client ->
+        client.reuseAddress = true
+        client.connect(
+          InetSocketAddress(
+            groupInfo.groupOwnerAddress.hostAddress,
+            PEER_HANDSHAKE_PORT
+          ), 15000
+        )
+        val objectOutputStream = ObjectOutputStream(client.getOutputStream())
+        // Send message for the peer device to verify
+        objectOutputStream.writeObject(HANDSHAKE_MESSAGE)
+        exchangeFileTransferMetadata(client.getInputStream(), client.getOutputStream())
+        groupInfo.groupOwnerAddress
+      }
+    } catch (ex: Exception) {
+      ex.printStackTrace()
+      null
+    }
+  }
+
+  private fun readHandshakeAndExchangeMetaData(): InetAddress? {
+    return try {
+      ServerSocket(PEER_HANDSHAKE_PORT)
+        .use { serverSocket ->
+          serverSocket.reuseAddress = true
+          val server = serverSocket.accept()
+          val objectInputStream = ObjectInputStream(server.getInputStream())
+          val kiwixHandShakeMessage = objectInputStream.readObject()
+
+          // Verify that the peer trying to communicate is a kiwix app intending to transfer files
+          if (isKiwixHandshake(kiwixHandShakeMessage)) {
+            if (BuildConfig.DEBUG) {
+              Log.d(TAG, "Client IP address: " + server.inetAddress)
+            }
+            exchangeFileTransferMetadata(server.getInputStream(), server.getOutputStream())
+            server.inetAddress
+          } else {
+            // Selected device is not accepting wifi direct connections through the kiwix app
+            null
+          }
+        }
+    } catch (ex: Exception) {
+      ex.printStackTrace()
+      return null
```

Comment:
the return can be omitted here. Is there something you would like explained to you on the nature of function flow? By all means ask questions.

Reply:
I'm not sure I understand the question. Can you give me some more details?

Reply:
Well I have in this pr, multiple times, requested returning try, avoiding labelled returns and requested the deleting of extraneous returns. I feel like the lesson of simplicity and brevity is not being imparted as the pattern persists. If you have an area of confusion or are not sure what is better or why it is better then questions are welcome

Reply:
I had some confusion about scope returns but you cleared that Thanks! I thought return@something if we don't that then it will doesn't returns.

Reply:
good, well get rid of this return as just saying `null` will do and I think we are there

Reply:
D'oh! I thought I removed all return keyword.